README.md
# Nuxt Smart App Banner [nuxt ^3.0.0]

[![npm version][npm-version-src]][npm-version-href]
[![npm downloads][npm-downloads-src]][npm-downloads-href]
[![License][license-src]][license-href]

![Alt text](android_view.png?raw=true "dsdfsdfsdfsdf") ![Alt text](ios_view.png?raw=true "Title")

Based on\
'jQuery Smart Banner' by Arnold Daniels <arnold@jasny.net> https://github.com/jasny/jquery.smartbanner ,\
'Smart App Banner' by Vitaly Glibin <glibin.v@gmail.com> https://github.com/kudago/smart-app-banner

> Smart App Banner for Nuxt

- [✨ &nbsp;Release Notes](/CHANGELOG.md)
<!-- - [📖 &nbsp;Documentation](https://example.com) -->

## Features

<!-- Highlight some of the features your module provide here -->
- Smart App Banner for Anroid/IOS
- Nuxt 3 compatible
- Support native IOS banner
- Callbacks available

## Quick Setup

1. Add `nuxt-smart-app-banner` dependency to your project

```bash
# Using pnpm
pnpm add -D nuxt-smart-app-banner

# Using yarn
yarn add --dev nuxt-smart-app-banner

# Using npm
npm install --save-dev nuxt-smart-app-banner
```

2. Add `nuxt-smart-app-banner` to the `modules` section of `nuxt.config.ts` and provide module config

```js
import { defineNuxtConfig } from 'nuxt/config'

export default defineNuxtConfig({
    modules: ['nuxt-smart-app-banner'],
    "nuxt-smart-app-banner": {
      bannerOptions: {
        title: "Title",
        androidAppId: "",
        icon: "https://icon-library.com/images/play-store-icon/play-store-icon-9.jpg",
        iosAppId: "",
        appStoreLanguage: "us",
      }
    }
})

```

3. Place smart banner component into desired component, page or layout\
⚠️ Note: if you set overrideComponentNameWith, use custom name instead of SmartAppBanner ⚠️
```html
<template>
  <div>
  🔥<SmartAppBanner/>🔥
    <NuxtWelcome />
  </div>
</template>
```

4. Using emits(callbacks)
```js
<template>
  <div>
    Nuxt module playground!
  </div>
  <SmartAppBanner 
    @on-dismiss="handleOnDismiss" // callback fired on clicking close button
    @on-install="handleOnInstall" // callback fired on clicking install button
    @on-not-shown="handleOnNotShown" // callback fired on banner not shown because it was previously either clicked or dismissed
    @on-shown="handleOnShown" // callback fired on banner shown
  />
</template>

<script setup>

const handleOnDismiss = (platform, appId) => {
  console.log(`handleOnDismiss ${platform} ${appId}`)
}
const handleOnInstall = (platform, appId) => {
  console.log(`handleOnInstall ${platform} ${appId}`)
}
const handleOnNotShown = (platform, appId, reason) => {
  console.log(`handleOnNotShown ${platform} ${appId} ${reason}`)
}
const handleOnShown = (platform, appId) => {
  console.log(`handleOnShown ${platform} ${appId}`)
}

</script>
```
That's it! You can now use Nuxt Smart App Banner in your Nuxt app ✨

## Config overview
```js
export interface ModuleOptions {
  overrideComponentNameWith?: string; // Use to override default component name
  bannerOptions: {
    useNativeIosBannerForSafari: boolean; // use native ios banner if available instead of custom (default: true)
    daysHidden: number;  // days to banner appear again after close button is clicked (defaults to 15)
    daysReminder: number;  // days to banner appear again after "VIEW" button is clicked (defaults to 90)
    appStoreLanguage: string; // ios app store language
    title: string;  // title text shown on banner
    author: string;  // author text shown on banner
    button: string;  // button text shown on banner
    androidAppId: string;  // android app id
    iosAppId: string; // ios app id
    store: {
        ios: string; // Instore text shown on banner for ios
        android: string; // Instore text shown on banner for android
    },
    price: {
        ios: string; // Price text shown on banner for ios
        android: string; // Price text shown on banner for ios
    },
    icons?: {
        ios: string; // Url for app icon shown on banner for ios
        android: string; // Url for app icon shown on banner for android
    },
    icon: string; // fallback icon url for all platforms
    theme?: SmartAppBannerTheme; // put platform type here to force single theme on all device
    force?: SmartAppBannerPlatform, // put platform type here for force banner platform for debug 
    }
}
```

## Development

```bash
# Install dependencies
npm install

# Generate type stubs
npm run dev:prepare

# Develop with the playground
npm run dev

# Build the playground
npm run dev:build

# Run ESLint
npm run lint
```

<!-- Badges -->
[npm-version-src]: https://img.shields.io/npm/v/nuxt-smart-app-banner/latest.svg?style=flat&colorA=18181B&colorB=28CF8D
[npm-version-href]: https://npmjs.com/package/nuxt-smart-app-banner

[npm-downloads-src]: https://img.shields.io/npm/dt/nuxt-smart-app-banner.svg?style=flat&colorA=18181B&colorB=28CF8D
[npm-downloads-href]: https://npmjs.com/package/nuxt-smart-app-banner

[license-src]: https://img.shields.io/npm/l/nuxt-smart-app-banner.svg?style=flat&colorA=18181B&colorB=28CF8D
[license-href]: https://npmjs.com/package/nuxt-smart-app-banner



CHANGELOG.md
..............................................................................................
## 0.2.0

### ❗Breaking changes
- removed windows mobile support
- module config changed

### Features
- added `useNativeIosBannerForSafari` config prop (default:true), if set use native ios banner when available
- added `icons` config prop used for separate icons by platform
- added `onInstall`, `onDismiss`, `onShown`, `onNotShown` callbacks
- added tests, lint
- optimise bundle size
- fix bugs

..............................................................................................
## 0.1.0 
Inital release



test/shared/selectors.ts
export const smartAppBannerElementsSelector = {
    viewButtonText: ".smartbanner-button-text",
    title: ".smartbanner-title",
    author: ".smartbanner-author",
    instore:".smartbanner-instore",
    viewButton: ".smartbanner-button",
    icon: ".smartbanner-icon",
    rootElement: {
        "android": ".smartbanner-android",
        "ios": ".smartbanner-ios"
    }
}

test/behavior/nuxt.config.ts
import MyModule from '../../src/module';
//import { SmartAppBannerNotShownReason, SmartAppBannerPlatform } from '../../src/runtime/types';

export default defineNuxtConfig({
  modules: [
    MyModule
  ],
  "nuxt-smart-app-banner": {
    bannerOptions: {
      title: "Title",
      androidAppId: "com.redlinegames.attackhole",
      icon: "https://icon-library.com/images/play-store-icon/play-store-icon-9.jpg",
      //onInstall: (platform: SmartAppBannerPlatform, appId: string) => console.log("onInstall click"),
      //onDismiss: (platform: SmartAppBannerPlatform, appId: string) => console.log("onDismiss click"),
      //onShown: (platform: SmartAppBannerPlatform, appId: string) => console.log("onShown click"),
      //onNotShown: (platform: SmartAppBannerPlatform, appId: string, reason: SmartAppBannerNotShownReason) => console.log(`onNotShown click ${reason}`),
    }
  }
})

test/platforms/nuxt.config.ts
import MyModule from '../../src/module';

export default defineNuxtConfig({
  modules: [
    MyModule
  ],
  "nuxt-smart-app-banner": {
    overrideComponentNameWith: '',
    bannerOptions: {
      useNativeIosBannerForSafari: true,
      daysHidden: 15,   // days to hide banner after close button is clicked (defaults to 15)
      daysReminder: 90, // days to hide banner after "VIEW" button is clicked (defaults to 90)
      title: 'Title',
      author: 'Company LLC',
      button: 'View',
      icon: '',
      icons: {
        "android": "https://icon-library.com/images/play-store-icon/play-store-icon-9.jpg",
        "ios": "https://icon-library.com/images/play-store-icon/play-store-icon-9.jpg",
      },
      store: {
        ios: 'On the App Store',
        android: 'In Google Play',
      },
      price: {
        ios: 'FREE',
        android: 'FREE anroid',
      },
      androidAppId: "com.redlinegames.attackhole",
      iosAppId: "426415753",
      appStoreLanguage: "by",
      // , theme: '' // put platform type ('ios', 'android', etc.) here to force single theme on all device
    }
  }
})

src/runtime/platformHelper.ts

import { Parser } from "bowser";
import { SmartAppBannerPlatform, SmartBannerOptions } from "./types";

interface PlatformData {
    platform: SmartAppBannerPlatform;
    iconRels: string[];
    getStoreLink: (bannerConfig: SmartBannerOptions) => string;
    getAppId: (bannerConfig: SmartBannerOptions) => string;
    getIcon: (bannerConfig: SmartBannerOptions) => string;
}

enum PlatformOs {
    'Android' = 'android',
    'iOS' = 'ios'
}

type PlatformDataKey = SmartAppBannerPlatform | PlatformOs

type PlatformsData = Record<PlatformDataKey, PlatformData>;

export const platformsData: PlatformsData = {
    ios: {
        platform: SmartAppBannerPlatform.ios,
        iconRels: ['apple-touch-icon-precomposed', 'apple-touch-icon'],
        getStoreLink: (bannerConfig: SmartBannerOptions) =>
            'https://itunes.apple.com/' + bannerConfig.appStoreLanguage + '/app/id' + bannerConfig.iosAppId + "?mt=8",
        getAppId: (bannerConfig: SmartBannerOptions) => bannerConfig.iosAppId,
        getIcon: (bannerConfig: SmartBannerOptions) => bannerConfig.icons?.ios || bannerConfig.icon,
    },
    android: {
        platform: SmartAppBannerPlatform.android,
        iconRels: ['android-touch-icon', 'apple-touch-icon-precomposed', 'apple-touch-icon'],
        getStoreLink: (bannerConfig: SmartBannerOptions) =>
            'http://play.google.com/store/apps/details?id=' + bannerConfig.androidAppId,
        getAppId: (bannerConfig: SmartBannerOptions) => bannerConfig.androidAppId,
        getIcon: (bannerConfig: SmartBannerOptions) => bannerConfig.icons?.android || bannerConfig.icon,
    }
}


export const identifyPlatform = function (bannerConfig: any, osName: string): SmartAppBannerPlatform | null {
    if (bannerConfig.force) {
        return bannerConfig.force;
    }
    if (Object.keys(PlatformOs).every(key => key !== osName)) {
        return null;
    }
    return platformsData[PlatformOs[osName]].platform;
}

export const isMobileSafariPlatform = function (computedTheme, bowser: Parser.Parser) {
    return (computedTheme === 'ios' && (bowser.getBrowserName() === 'Mobile Safari' || (bowser.getBrowserName() === 'Safari' && bowser.getPlatformType() === 'mobile')) && parseInt(bowser.getOSVersion(), 10) >= 6);
}


src/runtime/composables/useSmartAppBanner.ts
import { ref, computed, ComputedRef, Ref } from 'vue';
import { SmartAppBannerNotShownReason, SmartAppBannerPlatform, SmartBannerOptions } from '../types';
import { identifyPlatform, isMobileSafariPlatform, platformsData } from '../platformHelper';
import Cookies from 'js-cookie'
import Bowser from "bowser";
import { useHead, useRuntimeConfig } from '#imports'

const useSmartAppBannerCallbacks = (appId: ComputedRef<string>, platform: Ref<SmartAppBannerPlatform>, setShowBanner: (value: boolean) => void, emit: any, bannerConfig: SmartBannerOptions) => {

    const installClick = () => {
        setShowBanner(false);
        Cookies.set(appId.value + '-smartbanner-installed', 'true', {
            path: '/',
            expires: new Date(Number(new Date()) + (bannerConfig.daysReminder * 1000 * 60 * 60 * 24))
        });
        emit('onInstall', platform.value, appId.value);
    };

    const dismissClick = () => {
        setShowBanner(false);
        Cookies.set(appId.value + '-smartbanner-closed', 'true', {
            path: '/',
            expires: new Date(Number(new Date()) + (bannerConfig.daysHidden * 1000 * 60 * 60 * 24))
        });
        emit('onDismiss', platform.value, appId.value);
    };

    return { dismissClick, installClick }
}

const useSmartAppBannerState = (bannerConfig: SmartBannerOptions) => {
    const showBanner = ref<boolean>(false);
    const setShowBanner = (value: boolean) => {
        showBanner.value = value;
    }
    const platform = ref<SmartAppBannerPlatform>(SmartAppBannerPlatform.android);
    const setPlatform = (value: SmartAppBannerPlatform) => {
        platform.value = value;
    }
    const appId = computed(() => platformsData[platform.value].getAppId(bannerConfig));
    const storeLink = computed(() => platformsData[platform.value].getStoreLink(bannerConfig));
    const theme = computed(() => bannerConfig.theme || platform.value);
    const inStoreText = computed(() => bannerConfig.price[platform.value] + ' - ' + bannerConfig.store[platform.value]);
    const icon = computed(() => {
        const iconFromOptions = platformsData[platform.value].getIcon(bannerConfig);
        if (iconFromOptions) {
            return iconFromOptions;
        } else {
            const iconRels = platformsData[platform.value].iconRels;
            for (let i = 0; i < iconRels.length; i++) {
                const rel = window?.document.querySelector('link[rel="' + iconRels[i] + '"]');

                if (rel) {
                    return rel.getAttribute('href');
                }
            }
            return "";
        }
    });

    const mainContainerClass = computed(() => `smartbanner smartbanner-${theme.value}`);
    const iconStyle = computed(() => `background-image: url(${icon.value})`);

    return { showBanner, platform, appId, storeLink, theme, inStoreText, icon, setPlatform, setShowBanner, mainContainerClass, iconStyle };
}

export const useSmartAppBanner = (emit: any) => {
    const bannerConfig = useRuntimeConfig().smartAppBanner as SmartBannerOptions;
    const {
        appId,
        inStoreText,
        platform,
        setPlatform,
        setShowBanner,
        showBanner,
        storeLink,
        iconStyle,
        mainContainerClass,
    } = useSmartAppBannerState(bannerConfig);

    const { dismissClick, installClick } = useSmartAppBannerCallbacks(appId, platform, setShowBanner, emit, bannerConfig);

    const state = {
        title: bannerConfig.title, author: bannerConfig.author, buttonText: bannerConfig.button,
        showBanner, storeLink, inStoreText, dismissClick, installClick, iconStyle, mainContainerClass
    };

    const bowser = Bowser.getParser(window?.navigator?.userAgent);
    const platformValue = identifyPlatform(bannerConfig, bowser.getOSName());
    if (!platformValue) {
        return state;
    }

    setPlatform(platformValue);

    if (bannerConfig.useNativeIosBannerForSafari) {
        useHead({
            meta: [
                { name: 'apple-itunes-app', content: `app-id=${bannerConfig.iosAppId}` },
            ]
        })
    }

    // Don't show banner on ANY of the following conditions:
    // - device os is not supported,
    // - user is on mobile safari for ios 6 or greater (iOS >= 6 has native support for SmartAppBanner)
    // - running on standalone mode
    // - user dismissed banner
    const unsupported = !bannerConfig.store[platform.value];
    if (unsupported) {
        return state;
    }

    const isMobileSafari = bannerConfig.useNativeIosBannerForSafari && isMobileSafariPlatform(platform.value, bowser);
    const runningStandAlone = (window?.navigator as any).standalone;

    if (isMobileSafari || runningStandAlone) {
        return state;
    }

    const userDismissed = Cookies.get(appId.value + '-smartbanner-closed');
    const userInstalled = Cookies.get(appId.value + '-smartbanner-installed');

    if (userDismissed) {
        emit('onNotShown', platform.value, appId.value, SmartAppBannerNotShownReason.dismissed);
        return state;
    }

    if (userInstalled) {
        emit('onNotShown', platform.value, appId.value, SmartAppBannerNotShownReason.installed);
        return state;
    }

    // - If we dont have app id, dont display the banner
    if (!appId.value) {
        return state;
    }

    setShowBanner(true);
    emit('onShown', platform.value, appId.value);

    return state;
}

src/module.ts
import { fileURLToPath } from 'url'
import { defineNuxtModule, createResolver, addComponent } from '@nuxt/kit'
import { SmartBannerOptions } from "./runtime/types";
export interface ModuleOptions {
  overrideComponentNameWith?: string;
  bannerOptions: SmartBannerOptions;
}

export default defineNuxtModule<ModuleOptions>({
  meta: {
    name: 'nuxt-smart-app-banner',
    configKey: 'nuxt-smart-app-banner',
    compatibility: {
      nuxt: '^3.0.0'
    }
  },
  defaults: {
    bannerOptions: {
      useNativeIosBannerForSafari: true,
      daysHidden: 15,   
      daysReminder: 90, 
      appStoreLanguage: 'us',
      title: 'Title',
      author: 'Company LLC',
      button: 'View',
      icon: '',
      store: {
        ios: 'On the App Store',
        android: 'In Google Play',
      },
      price: {
        ios: 'FREE',
        android: 'FREE',
      },
      androidAppId: "",
      iosAppId: "",
    }

  },
  setup(options, nuxt) {
    const { resolve } = createResolver(import.meta.url)
    const runtimeDir = fileURLToPath(new URL('./runtime', import.meta.url))
    nuxt.options.build.transpile.push(runtimeDir)
    addComponent({
      name: options.overrideComponentNameWith || 'SmartAppBanner',
      global: true,
      filePath: resolve('./runtime/components/SmartAppBanner.vue'),
      mode: 'client'
    });

    nuxt.options.runtimeConfig.public.smartAppBanner = options.bannerOptions;
  }
})
